rfcs/45-stageless.md

```diff
@@ -2,488 +2,457 @@
-    // 3. CommandFlush::EndOfFrame (runs after data is extracted for rendering, at the end of the schedule)
-    .add_system(spawn_projectiles.before(CommandFlush::PostUpdate))
-    .add_systems(
-      (check_if_projectiles_hit, despawn_projectiles_that_hit)
-      .chain()
-      .after(CommandFlush::PostUpdate)
-    )
-    // If we need to add more command flushes to make our logic work, we can manually insert them
-    .add_system(flush_commands.in_set("DespawningFlush").after(CommandFlush::PostUpdate).before(CommandFlush::EndOfFrame))
-    .add_system(fork_projectiles.after("DespawningFlush"));
-  }
+```rust
+fn main() {
+    App::new()
+        /* ... */
+        .add_system(load_map.in_set(OnEnter(GameState::Playing)))
+        .add_system(autosave.in_set(OnExit(GameState::Playing)))
+        /* ... */
+        .run();
 }
 ```
 
-### Complex control flow
-
-Occasionally, you may find yourself yearning for more complex system control flow than "every system runs once in a loop".
-When that happens: **create an exclusive system and run a schedule in it.**
-
-Within an exclusive system, you can freely fetch the desired schedule from the `App` with the `&Schedules` system parameter and use `Schedule::run(&mut world)`, applying each of the systems in that schedule a single time to the world of the exclusive system.
-However, because you're in an ordinary Rust function you're free to use whatever logic and control flow you desire: branch and loop in whatever convoluted fashion you need!
-
-This can be helpful when:
-
-- you want to run multiple steps of a set of game logic during a single frame (as you might see in complex turn-based games)
-- you need to branch the entire schedule to handle a particularly complex bit of game logic
-- you need to integrate a complex external service, like scripting or a web server
-- you want to switch the executor used for some portion of your systems
-
-If you need absolute control over the entire schedule, consider having a single root-level exclusive system.
-
-#### Fixed time steps
+You can queue a state transition by getting a mutable reference to the state machine resource (`ResMut<Fsm<S>>`) and then using the `Fsm::queue_transition` method within your system.
 
-Running a set of systems (typically physics) according to a fixed time step is a particularly critical case of this complex control flow.
-These systems should run a fixed number of times for each second elapsed.
-The number of times these systems should be run varies (it may be 0, 1 or more each frame, depending on the time that the frame took to complete).
+Since state transitions can be as open-ended as commands, they're deferred (just like commands) until the next execution of an `apply_state_transition<S: State>` exclusive system. This system first extracts the FSM using `World::resource_scope`, next runs the "on-exit" schedule of the current state, then runs the "on-enter" schedule of the queued state, and finally updates the FSM's state.

```

Comment:
```suggestion
Since state transitions can be as open-ended as commands, they're deferred in a similar fashion until the next execution of an `apply_state_transition<S: State>` exclusive system.
This system first extracts the FSM using `World::resource_scope`, next runs the "on-exit" schedule of the current state, then runs the "on-enter" schedule of the queued state, and finally updates the FSM's state.
```